src/pudl/transform/eia860.py

```diff
@@ -603,25 +606,154 @@ def boilers(eia860_dfs, eia860_transformed_dfs):
     """
     # Populating the 'boilers_eia860' table
     b_df = eia860_dfs["boiler_info"].copy()
+    ecs = eia860_dfs["emission_control_strategies"].copy()
 
-    # Replace empty strings, whitespace, and '.' fields with real NA values
-    b_df = pudl.helpers.fix_eia_na(b_df)
+    # Combine and replace empty strings, whitespace, and '.' fields with real NA values
+
+    b_df = (
+        pd.concat([b_df, ecs], sort=True)
+        .dropna(subset=["boiler_id", "plant_id_eia"])
+        .pipe(pudl.helpers.fix_eia_na)
+    )
+
+    # Replace 0's with NaN for certain columns.
+    zero_columns_to_fix = [
+        "firing_rate_using_coal_tons_per_hour",
+        "firing_rate_using_gas_mcf_per_hour",
+        "firing_rate_using_oil_bbls_per_hour",
+        "firing_rate_using_other_fuels",
+        "fly_ash_reinjection",
+        "hrsg",
+        "new_source_review",
+        "turndown_ratio",
+        "waste_heat_input_mmbtu_per_hour",
+    ]
 
+    for column in zero_columns_to_fix:
+        b_df[column] = b_df[column].replace(to_replace=0, value=np.nan)
+
+    # Fix unlikely year values for compliance year columns
+    year_cols_to_fix = [
+        "compliance_year_nox",
+        "compliance_year_so2",
+        "compliance_year_mercury",
+        "compliance_year_pm",
+    ]
+
+    for col in year_cols_to_fix:
+        b_df.loc[b_df[col] < 1900, col] = pd.NA
+
+    # Convert boolean columns from Y/N to True/False.
     boolean_columns_to_fix = [
         "hrsg",
         "fly_ash_reinjection",
         "new_source_review",
+        "mercury_emission_control",
+        "ACI",
+        "BH",
+        "DS",
+        "EP",
+        "FGD",
+        "LIJ",
+        "WS",
+        "BS",
+        "BP",
+        "BR",
+        "EC",
+        "EH",
+        "EK",
+        "EW",
+        "OT",
     ]
 
     for column in boolean_columns_to_fix:
         b_df[column] = (
             b_df[column]
             .fillna("NaN")
-            .replace(to_replace=["Y", "N", "NaN"], value=[True, False, pd.NA])
+            .replace(
+                to_replace=["Y", "N", "NaN", "0"], value=[True, False, pd.NA, pd.NA]
+            )
         )
 
+    # 2009-2012 data uses boolean columns for mercury equipment that
+    # later are converted to strategy codes. Here we convert them manually.
+
+    mercury_boolean_cols = [
+        "ACI",
+        "BH",
+        "DS",
+        "EP",
+        "FGD",
+        "LIJ",
+        "WS",
+        "BS",
+        "BP",
+        "BR",
+        "EC",
+        "EH",
+        "EK",
+        "EW",
+        "OT",
+    ]
```

Comment:
Can you expand on the comment here explaining what's going on with the mercury boolean columns? It sounds like there was a structural change in how the old vs. new data is being reported, and you're reshaping the old data to conform to the new structure, which is N columns that contain values from the same set of codes.

Not something to tackle right now, but it looks like there are a growing number of collections of numbered coded columns like this, and I wonder if we might want to think about how to best normalize them in the DB. Having a single column that can contain one of several different codes makes sense (e.g. with prime mover), but when we need to be able to record up to 6 or 8 different values, I wonder if boolean columns, with one per potential code, might make more sense. Or we could do a bad thing and put a list of codes in a single column.  Is there information in the cardinality of the columns? I.e. is the emissions control strategy in the `_1` column more important than the one in the `_2` column?

How do we imagine folks using these columns in analysis? What structure will make selecting on the basis of the codes present easiest? If you wanted to use them as features for an ML algorithm (say, to estimate the expected emissions for a unit that doesn't report its emissions directly in the EPA CEMS) how would you want to arrange them? Is it ever possible to have more than 6 mercury control strategies on a single unit?

Reply:
Yes, I agree. I think the best course of action will be clearer once we decide how we want to integrate the respective equipment data columns. It could be that we just replace all of these columns with a link to the respective equipment IDs and keep information about equipment types in those dataframes, or that we want to pursue some kind of boolean reorganization. Currently there are loads of columns that differentiate between different contexts (current/planned, what regulation is being met, etc.), so we could also pursue recoding in other columns (e.g. most string applicable regulation) to reduce the number of columns functioning here. There is no information about cardinality in the columns explicitly specified in the layout files or the survey, so I don't think we can presume that here.